src/include/duckdb/parallel/pipeline_executor.hpp

```diff
@@ -32,6 +32,25 @@ enum class PipelineExecuteResult {
 	INTERRUPTED
 };
 
+class ExecutionBudget {
+public:
+	explicit ExecutionBudget(idx_t maximum) : processed(0), maximum_to_process(maximum) {
+	}
+
+public:
+	bool Next() {
+		if (processed >= maximum_to_process) {
+			return false;
+		}
+		processed++;
+		return true;
+	}
```

Comment:
Can we add this method here?
```c++
bool IsDepleted() const {
    return processed >= maximum_to_process;
}
```
And use it in the `Next()` function (and possibly some assertions as well?).

Also, should `processed` ever be greater than `maximum_to_process`? Maybe we should assert that it is less or equal in `IsDepleted`.